src/virtual_machine/lc3_virtual_machine.rs

```diff
@@ -0,0 +1,384 @@
+use std::io::Read;
+
+use super::{instructions::*, register::Register, trap::Trap};
+
+const AMMOUNT_OF_REGISTERS: usize = 10;
+pub enum Flag {
+    POSITIVE = 1 << 0,
+    ZERO = 1 << 1,
+    NEGATIVE = 1 << 2,
+}
+
+enum MemoryMappedRegister {
+    KeyBoardStatusRegister = 0xFE00,
+    KeyBoardDataRegister = 0xFE02,
+}
+
+pub struct LC3VirtualMachine {
+    registers: Vec<u16>,
+    memory: Vec<u16>,
+}
+
+impl LC3VirtualMachine {
+    pub fn new(program_counter_start: u16) -> Self {
+        let mut registers = vec![0; AMMOUNT_OF_REGISTERS];
+        registers[Register::ProgramCounter as usize] = program_counter_start;
+        Self {
+            registers,
+            memory: vec![0; 1 << 16],
+        }
+    }
+
+    pub fn read_register(&self, source_register: Register) -> u16 {
+        self.registers[source_register as usize]
+    }
+
+    pub fn update_register(&mut self, destination_register: Register, new_register_value: u16) {
+        self.registers[destination_register as usize] = new_register_value;
+    }
+
+    pub fn update_flags(&mut self, register: Register) {
+        let register = register as usize;
+        if self.registers[register] == 0 {
+            self.update_register(Register::ConditionFlag, Flag::ZERO as u16)
+        } else if (self.registers[register] >> 15) != 0 {
+            self.update_register(Register::ConditionFlag, Flag::NEGATIVE as u16)
+        } else {
+            self.update_register(Register::ConditionFlag, Flag::POSITIVE as u16)
+        }
+    }
+
+    fn receive_keyboard_input(&mut self) {
+        let mut buffer = [0; 1];
+        std::io::stdin().read_exact(&mut buffer).unwrap();
+        if buffer[0] != 0 {
+            self.memory_write(MemoryMappedRegister::KeyBoardStatusRegister as u16, 1 << 15);
+            self.memory_write(
+                MemoryMappedRegister::KeyBoardDataRegister as u16,
+                buffer[0] as u16,
+            );
+        } else {
+            self.memory_write(MemoryMappedRegister::KeyBoardStatusRegister as u16, 0)
+        }
+    }
+
+    pub fn memory_read(&mut self, memory_address: u16) -> u16 {
+        if memory_address == MemoryMappedRegister::KeyBoardStatusRegister as u16 {
+            self.receive_keyboard_input();
+        }
+        self.memory[memory_address as usize]
+    }
+
+    pub fn memory_write(&mut self, memory_address: u16, value_to_write: u16) {
+        self.memory[memory_address as usize] = value_to_write;
+    }
+
+    pub fn process_input(&mut self, instruction: u16) {
+        let instruction_opcode = Instruction::from(instruction >> 12);
+        match instruction_opcode {
```

Comment:
Is there a reason for not including the instruction construction directly in the corresponding functions? I believe it would be cleaner and easier to test.

Reply:
I definitely support this.

Reply:
It was that way at first, but after thinking about it, I don't feel it makes sense that the instructions should have that responsability, as mentioned in another comment, the VM should decode and decide what instruction to execute, the instructions do the rest